modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,371 @@
+	u := &Tracker{
+		cfg:     cfg,
+		name:    name,
+		tenants: make(map[string]*tenantUsage),
+		fn:      fn,
+		reg:     prometheus.NewRegistry(),
+	}
+
+	err := u.reg.Register(u)
+	if err != nil {
+		return nil, err
+	}
+
+	go u.PurgeRoutine()
+
+	return u, nil
+}
+
+// getTenant must be called under lock.
+func (u *Tracker) getTenant(tenant string) *tenantUsage {
+	data := u.tenants[tenant]
+	if data == nil {
+		data = &tenantUsage{
+			series: make(map[uint64]*bucket),
+			constLabels: prometheus.Labels{
+				tenantLabel:  tenant,
+				trackerLabel: u.name,
+			},
+		}
+		u.tenants[tenant] = data
+	}
+	return data
+}
+
+func (u *Tracker) Observe(tenant string, batches []*v1.ResourceSpans) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	dimensions := u.fn(tenant)
+	if len(dimensions) == 0 {
+		// Not configured
+		// TODO - Should we put it all in the unattributed bucket instead?
+		return
+	}
+
+	var (
+		now            = time.Now().Unix()
+		data           = u.getTenant(tenant)
+		labels, values = data.GetBuffersForDimensions(dimensions)
+	)
+
+	for _, batch := range batches {
+		unaccountedForBatchData, totalSpanCount := nonSpanDataLength(batch)
+
+		if totalSpanCount == 0 {
+			// Mainly to prevent a panic below, but is this even possible?
+			continue
+		}
+
+		// This is 1/Nth of the unaccounted for batch data that gets added to each span.
+		// Adding this incrementally as we go through the spans is the fastest method, but
+		// loses some precision. The other (original) implementation is to record span counts
+		// per series into a map and reconcile at the end. That method has more accurate data because
+		// it performs the floating point math once on the total, instead of accumulating 1/N + 1/N ... errors.
+		batchPortion := int(math.RoundToEven(float64(unaccountedForBatchData) / float64(totalSpanCount)))
+
+		// To account for the accumulated error we dump the remaining delta onto the first span, which can be negative.
+		// The result ensures the total recorded bytes matches the input.
+		firstSpanPortion := unaccountedForBatchData - batchPortion*totalSpanCount
+
+		// Reset value buffer to be empties
+		clear(values)
+
+		if batch.Resource != nil {
+			for _, a := range batch.Resource.Attributes {
+				v := a.Value.GetStringValue()
+				if v == "" {
+					continue
+				}
+				for i, d := range dimensions {
+					if d == a.Key {
+						values[i] = v
+					}
+				}
+			}
+		}
+
+		var (
+			bucket      *bucket
+			bucketDirty = false
+		)
+
+		for i, ss := range batch.ScopeSpans {
+			for j, s := range ss.Spans {
+				sz := s.Size()
+				sz += protoLengthMath(sz)
+				sz += batchPortion // Incrementally add 1/Nth worth of the unaccounted for batch data
+				if i == 0 && j == 0 {
+					sz += firstSpanPortion
+				}
+
+				for _, a := range s.Attributes {
+					v := a.Value.GetStringValue()
+					if v == "" {
+						continue
+					}
+					for i, d := range dimensions {
```

Comment:
so it looks like span data "overrides" resource level data. should we require scopes on the configured dimensions to enforce level?

I'm imagining a scenario like:
```
batches: [
  atts: {
    "foo": "A",
  },
  spans: [
    { 
      atts: {
        "foo": "B",
      },
    },
    {
       // no value for foo
    },
  ]
]
```

will the first span overwrite the foo value to be "B" and that will carry forward to the second span even though it doesn't have a value for "foo"?

Reply:
Definitely considered this, but for this use case we need to check both areas, as different applications can be instrumented differently (service A puts the attribute on the resource, and service B puts the attribute on the span).  Was thinking we could evolve into that in the future, parsing as a traceql identifier allows us to know if it's a plain old label (current), or scoped (span.foo) or even unscoped again (.foo).

Another consideration was to keep the configuration simple, and consistent across signals. For example here is the PR for Mimir: https://github.com/grafana/mimir/pull/9392

Ultimately I think at this time I wouldn't want to require scopes yet.

Reply:
Ah sorry reading closer, yes I think you are right and that is a bug.   It looks like "B" would carry forward to the second span.

Reply:
This is fixed and added a test case.   Did take a small regression in performance because basically it becomes harder to detect/reuse the last hashed series across spans.